### ISSUE TITLE
LAPACK calculated eigenvector signs differ between processes when running in parallel

### ISSUE THREAD

**Issue by robertrueger** [primme/primme#3]
When the small eigenvalue problem is solved in solve_H_d.c, the signs of the resulting eigenvectors are arbitrary. This is in itself not an issue, as the sign of an eigenvector is not well defined: if v is an eigenvector, so is -v. However, when running in parallel the signs of the eigenvectors can differ between the different processes. This is not a problem during the calculation, but if the complete eigenvectors are assembled in the end from the parts of the basis residing on the different processes, the sign of each part is essentially random, leading to wrong results.

Take as an example a 4x4 matrix for which an eigenvector is (a,b,c,d) or equivalently (-a,-b,-c,-d). Primme running with two processes might produce (a,b,-c,-d) as an eigenvector if the signs of eigenvectors _of the small eigenvalue problem_ do not match between processes due to some randomness of the underlying dsyev.

This does not seem to happen on most platforms and LAPACK libraries, but we have observed the issue on Win32 with the MKL library: Diagonalizing the same matrix gives eigenvectors with different signs for the different processes running primme in parallel.

Here are three workarounds from the top of my head:
1. Introduce an additional function pointer for the diagonalization instead of calling dsyev. The primme user can then set it and handle the problem outside of primme. This is what we are currently doing in ADF, where the dsyev call is replaced with ScaLAPACK (which ensures a consistent sign across all processes). Code for this workaround (only real matrices) can be found [here](https://github.com/SCM-NV/primme/commit/d0edc150c9be8ae3dd62a7ba9362d4c504edad15).
2. Communicate the eigenvectors from procID=0 to all other processes. This can be done (though not super efficiently) using the globalSumDouble function that has to be set for parallel runs anyway. Code for this workaround (only real matrices) can be found [here](https://github.com/SCM-NV/primme/commit/d20c8e6183e92cfca3a3377a755850ba3073758e).
3. Come up with some sign convention for the eigenvectors of the small eigenvalue problem. (Something like "largest element positive".) This probably works for most cases, but I have doubts if it can work in general: It seems like partitioning space into two parts, one of which would have the sign flipped, would always introduce some border at which the sign is random if there is even the slightest noise on the eigenvector ...

We currently use primme 1.2 with workaround 1. in ADF. It would however be nice if this problem could be solved upstream.

**Comment by primme** [comment]
Have you tried to activate the reproducibility flags in MKL? Just set 
the environment variable MKL_CBWR = COMPATIBLE.

**Comment by stathopoulos** [comment]
Hi Robert and thank you again. 
Yes, we are aware of this potential problem, and we have seen it on heterogeneous platforms in the past. However, even there the issue was very rare so it has taken back seat in our Things To Do. Especially since most of HPC platforms use identical architecture/libraries. 
I am really surprised that MKL does this. There must be some randomization seed there that is activated in certain cases. Do you notice this all the time or rarely? Also, do you notice it on 64bit windows?

My plan was to go with solution 2 because I did not want to be dependent on another lib like ScaLapack. Also, a call to scalapack for such small matrices could be detrimental to performance. A broadcase would be cheaper.

Since you experience this problem we will have to speed up its solution.

Best,
Andreas

**Comment by robertrueger** [comment]
It has been a while since we encountered this problem and implemented the workaround with ScaLAPACK. Must have been one or two years ago when primme v1.1 was the latest release. It broke our tests on only one platform, which was 32bit Windows with some version of the MKL as far as I recall. I think it didn't happen every time, but only sometimes. All other platforms were fine though (including Windows 64bit), so yes, it is indeed a very rare issue. It would be quite a lot of work to go back, set up the old environment to try to reproduce the issue. So I can not test if MKL_CBWR = COMPATIBLE would fix it.

I'm only reporting the issue now because I was looking into updating our primme version from 1.2 to 1.2.2, when I stumbled upon our workaround and thought this should probably be fixed upstream.

Best regards,
Robert

PS: You are right, the use of ScaLAPACK for the diagonalization is probably not too efficient. It was however super easy to do, and I think in practice the performance doesn't matter here, since everything else is just so much slower ...
